src/contexts/AuthContext.tsx

```diff
@@ -30,50 +35,88 @@ const AuthProvider: React.FC<Props> = ({children}: Props) => {
   const [isAuthenticated, setIsAuthenticated] = useState(false);
   const [isAdmin, setIsAdmin] = useState(false);
   const [isLoading, setIsLoading] = useState(false);
+  const [email, setEmail] = useState('');
+  const [name, setName] = useState('');
   // --------------- End state maintenance variables ------------------------
 
+  /**
+   * Function to reset all state variables to defaults
+   */
+  function resetState(): void {
+    setIsAdmin(false);
+    setEmail('');
+    setName('');
+  }
+
   useEffect(() => {
     setIsLoading(true);
+
+    // Creates listener for authentication status
     const unsubscribe = firebaseAuth.onAuthStateChanged(user => {
+      setIsLoading(true);
       if (user) {
         setIsAuthenticated(true);
-
-        // Check if user is an admin user
-        firestore
-          .collection('users')
-          .doc(user.uid)
-          .get()
-          .then(doc => {
-            if (doc.exists) {
-              const userData = doc.data();
-              if (userData) {
-                if (userData.admin) setIsAdmin(true);
-              }
-            }
-          })
-          .catch(error => {
-            // Error thrown upon failure to fetch the users doc from Firestore
-            throw new Error(`Unable to fetch users doc: ${error}`);
-          })
-          .finally(() => {
-            // Loading is only finished after the async calls to Firestore complete
-            setIsLoading(false);
-          });
       } else {
-        setIsAuthenticated(false);
-        setIsAdmin(false);
-        setIsLoading(false);
+        resetState();
       }
+      setIsLoading(false);
     });
     return unsubscribe;
   }, []);
 
+  // Watch a user's document for account updates if a user is signed in
+  useEffect(() => {
+    if (isAuthenticated && firebaseAuth.currentUser) {
+      setIsLoading(true);
+      const user = firebaseAuth.currentUser;
+
+      // This creates a listener for changes on the document so that if a user
+      // updates their account information, this change is reflected on the user's
+      // account page.
+      const unsubscribe = firestore
+        .collection('users')
+        .doc(user.uid)
+        .onSnapshot(snapshot => {
+          if (snapshot.exists) {
+            const userData = snapshot.data();
+            if (userData) {
+              if (userData.admin !== undefined) setIsAdmin(userData.admin);
```

Comment:
should we also check that userData.admin is a boolean to avoid throwing an error?

Reply:
Good point, I'll add some type safety stuff here